File: prob1.py
# Time Complexity : O(M*N) as we got through all matrix elements in 2 for loops
# Space Complexity : O(M*N) for the queue
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : made the mistake of popping the element out of the level
# Your code here along with comments explaining your approach

# implement bfs, first keep a count of the fresh oranges, 
# and add rotten oranges to queue
# if at this point, fresh=0, return 0 as no oranges there to rot
# now from queue, keep track of size as we need the level here
# in that level find neighboring oranges in all 4 directions and if they 
# are fresh, rot them , add to queue and reduce fresh by 1
# in the end if fresh > 0, return -1, else return (time-1) as we increment the time 
# one extra after the last level 


from collections import deque

class Solution:

    def orangesRotting(self, grid):

        if grid is None or len(grid) == 0:
            return 0
        m, n = len(grid), len(grid[0])

        fresh = 0
        time = 0
        q = deque()
            
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    fresh += 1
                if grid[i][j] == 2:
                    q.append([i, j])
            
        if fresh == 0:
            return 0
            
        dirs = [[-1, 0], [0, -1], [1, 0], [0, 1]]

        while len(q) != 0:
            size = len(q)
                
            for i in range(size):
                curr = q.popleft()
                for d in dirs:
                    r = curr[0] + d[0]
                    c = curr[1] + d[1]
                    if r >= 0 and c >= 0 and r < m and c < n and grid[r][c] == 1:
                        grid[r][c] = 2
                        fresh -= 1
                        q.append([r, c])
            time += 1
            
        if fresh > 0:
            return -1
        return (time -1)
        

File: prob1.java
// Time Complexity : O(M*N) | for DFS, O(M*N)^2
// Space Complexity : O(M*N) 
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : No

// Explain your approach: this problem can be done using bith BFS and DFS
// using BFS, we keep count of fresh oranges and we push all the rotten oranges into the queue
// keeping track of size / level, at every level, we check 4 neighbours of each rotten orange in queue
// if any of the in biund neightbours are fresh, we rot them. At the end of each level, we increase time
// in the end, if fresh count has become 0, we return time - 1, else return -1
// one edge case, before iterating over queue elements, if fresh count is 0, from there itself return 0
// it means there are no fresh oranges present to rot.


import java.util.*;
class Solution{
    public int rottenOranges(int[][] grid){
        if(grid == null || grid.length == 0) return 0;
        int m = grid.length; int n = grid[0].length;
        Queue<int[]> q = new LinkedList<>();
        int fresh = 0;
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(grid[i][j] == 1) fresh++;
                else if(grid[i][j] == 2){
                    q.add(new int[]{i, j});
                }
            }
        }
        int time = 0;
        if(fresh == 0) return 0;
        int[][] dirs = {{-1, 0}, {0, -1}, {0, 1}, {1, 0}};
        while(!q.isEmpty()){
            int size = q.size();
            for(int i = 0; i < size; i++){
                int[] curr = q.poll();
                for(int[] dir: dirs){
                    int r = dir[0] + curr[0];
                    int c = dir[1] + curr[1];
                    if(r >= 0 && c >= 0 && r < m && c < n && grid[r][c] == 1){
                        grid[r][c] = 2;
                        q.add(new int[]{r, c});
                        fresh--;
                    }
                }

            }
            time++;
        }
        if(fresh > 0) return -1;
        return time - 1;
    }
}